HW2/P5/P5.txt

```diff
@@ -0,0 +1,86 @@
+1) Serial run yields performance of between 5 to 6 simulations per second, while parallel run yields 7-8 simulations per second. Therefore performance is better for 4 thread case, but not four times.
+
+2) Spatial decomposition surprisingly yields really good performance improvement from 7 to around 300 simulations per second. This is because a large part of calculation is done comparing whether two balls are close enough to each other to induce a collision. By decomposing space, the comparison's complexity is linear O(N) instead of a quadratic O(N^2) -  a significant increase and therefore boosting up the performance.
+
+3) Spatially coherent sorting improves performance from around 300 to around 360 simulations per second - a 20% increase. I thought for a while and used the following weighting grid to assign weights to each ball:
+ 1  2  9 10
+ 4  3  8 11
+ 5  6  7 12
+16 15 14 13...
+Then I googled and found a better weighting scheme - zorder curve, which is more balanced and efficient, but I think complexity is slightly higher than my scheme. The position list is then sorted according to the weights assigned.
+
+4) After adding locks, the performance decreased from 360 to around 320 simulations per second. Locking is necessary to prevent updating simutaneously. These safety checks impede performance by acquiring the locks and wait for potential collisions to pass and then release them, and they are essential for the sound physical movement of the colliding balls system.
+
+Through this problem I learned to use several ways to improve the performance of simple ball collision physics simulation: multithreading, decomposition and algorithm. The speed enhancement is significant. It also makes me think sometimes it would be better to first stop and think about the problem itself: is there any way to optimize the problem itself, rather than justing parallelizing it. It'll often surprise us how good this can perform with some simple optimization like the space decomposition we have in this problem.
+
```

Comment:
Would have been interesting to see your results for the various section also broken up by how many threads there were (i.e. for 1 thread and for four threads). For example, 4 threads & spatial decomposition, I got to around 500-680 fps. Was in-line with yours for the 1-thread case.